modules/ldapbackend/ldapauthenticator.cc

```diff
@@ -163,98 +190,109 @@ int LdapGssapiAuthenticator::attemptAuth( LDAP *conn )
+    krb5_kt_close( d_context, keytab );
+    return code;
   }
 
-  krb5_kt_end_seq_get( context, keytab, &cursor );
-  if ( code != 0 ) {
-    L<<Logger::Error << logPrefix << "krb5 error when extracting principal information: " << std::string( krb5_get_error_message( context, code ) ) << std::endl;
-    krb5_kt_close( context, keytab );
-    krb5_free_principal( context, principal );
+  if ( ( code = krb5_copy_principal( d_context, entry.principal, &principal ) ) != 0 ) {
+    L<<Logger::Error << d_logPrefix << "krb5 error when extracting principal information: " << std::string( krb5_get_error_message( d_context, code ) ) << std::endl;
+    krb5_kt_close( d_context, keytab );
+    krb5_kt_free_entry( d_context, &entry );
     return code;
   }
 
-  // Locate the credentials cache file
-  if ( !cCacheFile.empty() ) {
-    std::string cCacheStr( "FILE:" + cCacheFile );
-    code = krb5_cc_resolve( context, cCacheStr.c_str(), &ccache );
+  krb5_kt_free_entry( d_context, &entry );
+  krb5_kt_end_seq_get( d_context, keytab, &cursor );
+
+  if ( ( code = krb5_get_init_creds_opt_alloc( d_context, &options ) ) != 0 ) {
+    L<<Logger::Error << d_logPrefix << "krb5 error when allocating credentials cache structure: " << std::string( krb5_get_error_message( d_context, code ) ) << std::endl;
+    krb5_kt_close( d_context, keytab );
+    krb5_free_principal( d_context, principal );
+    return code;
   }
-  else {
-    code = krb5_cc_default( context, &ccache );
+  krb5_get_init_creds_opt_set_default_flags( d_context, "pdns", krb5_principal_get_realm( d_context, principal ), options );
+
+  // Get the ticket
+  code = krb5_get_init_creds_keytab( d_context, &credentials, principal, keytab, 0, NULL, options );
+  if ( code ) {
+    L<<Logger::Error << d_logPrefix << "krb5 error when getting the TGT: " << std::string( krb5_get_error_message( d_context, code ) ) << std::endl;
+    krb5_free_cred_contents( d_context, &credentials );
+    krb5_kt_close( d_context, keytab );
+    krb5_free_principal( d_context, principal );
+    return code;
   }
 
-  if ( code != 0 ) {
-    L<<Logger::Error << logPrefix << "krb5 error when locating the credentials cache file: " << std::string( krb5_get_error_message( context, code ) ) << std::endl;
-    krb5_kt_close( context, keytab );
-    krb5_free_principal( context, principal );
+  krb5_get_init_creds_opt_free( d_context, options );
+  krb5_kt_close( d_context, keytab );
+
+  // Use a temporary cache to get the initial credentials. This will be moved to the user-configured one later.
+  krb5_ccache tmp_ccache = NULL;
+
+  code = krb5_cc_new_unique( d_context, krb5_cc_get_type( d_context, d_ccache ), NULL, &tmp_ccache );
+  if ( code ) {
+    L<<Logger::Error<< d_logPrefix << "krb5 error when creating the temporary cache file: " << std::string( krb5_get_error_message( d_context, code ) ) << std::endl;
+    krb5_kt_close( d_context, keytab );
+    krb5_free_principal( d_context, principal );
```

Comment:
Aren't we missing a call to `krb5_free_cred_contents( d_context, &credentials );`?